storage/src/test/scala/coop/rchain/storage/StorageActionsTests.scala

```diff
@@ -0,0 +1,467 @@
+      store.getPs(txn, produceKey2) shouldBe Nil
+      store.getAs(txn, produceKey2) shouldBe List("datum2")
+      store.getK(txn, produceKey2) shouldBe None
+    }
+
+    r2 shouldBe None
+
+    val r3 = produce(store, produceKey3.head, "datum3")
+
+    store.withTxn(store.createTxnRead()) { txn =>
+      store.getKey(txn, produceKey3Hash) shouldBe produceKey3
+      store.getPs(txn, produceKey3) shouldBe Nil
+      store.getAs(txn, produceKey3) shouldBe List("datum3")
+      store.getK(txn, produceKey3) shouldBe None
+    }
+
+    r3 shouldBe None
+
+    val r4 = consume(store, List("ch1", "ch2", "ch3"), patterns, capture(results))
+
+    store.withTxn(store.createTxnRead()) { txn =>
+      store.getKey(txn, consumeKeyHash) shouldBe Nil
+      store.getPs(txn, consumeKey) shouldBe Nil
+      store.getAs(txn, consumeKey) shouldBe Nil
+      store.getK(txn, consumeKey) shouldBe None
+    }
+
+    r4 shouldBe defined
+
+    runK(r4)
+
+    results should contain theSameElementsAs List(List("datum1", "datum2", "datum3"))
+  }
+
+  "producing three times on the same channel then consuming three times on the same channel" should
+    "return three pairs of continuations and data" in withTestStore { store =>
+    val results: mutable.ListBuffer[List[String]] = mutable.ListBuffer.empty[List[String]]
+
+    val key = List("ch1")
+
+    val r1 = produce(store, key.head, "datum1")
+    val r2 = produce(store, key.head, "datum2")
+    val r3 = produce(store, key.head, "datum3")
+
+    r1 shouldBe None
+    r2 shouldBe None
+    r3 shouldBe None
+
+    val r4 = consume(store, key, List(Wildcard), capture(results))
+    val r5 = consume(store, key, List(Wildcard), capture(results))
+    val r6 = consume(store, key, List(Wildcard), capture(results))
+
+    store.withTxn(store.createTxnRead()) { txn =>
+      store.getKey(txn, store.hashC(key)) shouldBe key
+      store.getAs(txn, key) shouldBe Nil
+      store.getPs(txn, key) shouldBe Nil
+      store.getK(txn, key) shouldBe None
+    }
+
+    val continuations = List(r4, r5, r6)
+
+    continuations.forall(_.isDefined) shouldBe true
+
+    continuations.foreach(runK)
+
+    results should contain theSameElementsAs List(List("datum3"), List("datum2"), List("datum1"))
+  }
+
+  "consuming three times on the same channel, then producing three times on that channel" should
+    "return the last continuation consumed along with a piece of data" in withTestStore { store =>
+    val results: mutable.ListBuffer[List[String]] = mutable.ListBuffer.empty[List[String]]
+
+    consume(store, List("ch1"), List(Wildcard), capture(results))
+    consume(store, List("ch1"), List(Wildcard), capture(results))
+    consume(store, List("ch1"), List(Wildcard), capture(results))
+    val r1 = produce(store, "ch1", "datum1")
+    val r2 = produce(store, "ch1", "datum2")
+    val r3 = produce(store, "ch1", "datum3")
+
+    r1 shouldBe defined
+    r2 shouldBe None
+    r3 shouldBe None
+
```

Comment:
Yeah I think this test confirms what I asked above. I would expect r1, r2, and r3 to return one continuation each - the order is based on some kind of ranking algorithm. Having a ranking algorithm does slow down the tuplespace though - maybe we should just return the first consume candidate that we find for now (it also simplifies the code).

Reply:
In the RBL equivalent, I just used "take the head/first" as the ranking algorithm.

Reply:
Ah interesting, I had a conversation with @guardbotmk3 that left me with the impression that we would only store one continuation per key - is this incorrect?

Reply:
Isn't that was I proposed in discord a week or so back? I thought the solution to use DNS mappings or hashes as names would solve that issue.